### ISSUE TITLE
Questions about per_request_max_new_tokens

### ISSUE THREAD

**Issue by QsingHuan** [ShopeeLLM/Spec-RL#13]
Great work!
I have a question about the per_request_max_new_tokens item. 
I think that for a given prompt with draft response, the max_new_tokens should be reset for the inference engine to be max_new_tokens - draft_length. I do see that an item named per_request_max_new_tokens is calculated in the code, but I can't find where it is used. And I don't think that this is passed to control the inference engine. 
Thanks for your reply!

**Comment by bingshuailiu** [comment]
Hey @QsingHuan, great catch — and thanks a lot for reading the code so carefully! 🙏
Oops, my bad. When I uploaded this repo, the internal version had some sensitive files, so I re-uploaded only the core modified ones to GitHub. I did a quick end-to-end test after cloning, and everything seemed fine, but I totally forgot to include the updated max_tokens logic hidden inside the vLLM rollout file.
I’ve just pushed a new commit that re-uploads the vLLM rollout engine file.
 Now the per-request max_new_tokens is properly handled inside generate_sequences: when spec_decoding is True, it loops over each request’s max_tokens value to build the sampling params dynamically — see the relevant snippet below 👇
```python
spec_decoding = prompts.meta_info.get("spec_decoding", False)
with self.update_sampling_params(**kwargs):
    if spec_decoding:
        base_sp = deepcopy(self.sampling_params)

        per_req = prompts.non_tensor_batch['per_request_max_new_tokens']

        sampling_params_arg = base_sp
        if per_req is not None:
            if isinstance(per_req, torch.Tensor):
                per_req_list = per_req.tolist()
            else:
                per_req_list = list(per_req)
            assert len(per_req_list) == len(vllm_inputs), \
                f"per_request_max_new_tokens length {len(per_req_list)} != batch {len(vllm_inputs)}"

            sps = []
            for m in per_req_list:
                sp = deepcopy(base_sp)
                m_int = max(1, int(m))
                if hasattr(sp, "max_tokens"):
                    sp.max_tokens = m_int
                else:
                    sp.max_new_tokens = m_int
                sps.append(sp)
            sampling_params_arg = sps

            outputs = self.inference_engine.generate(
                prompts=vllm_inputs,
                sampling_params=sampling_params_arg,
                lora_request=lora_requests,
                use_tqdm=False,
            )
    else:
        outputs = self.inference_engine.generate(
            prompts=vllm_inputs,
            sampling_params=self.sampling_params,
            lora_request=lora_requests,
            use_tqdm=False,
        )
```
Please pull the latest commit and check again — it should now respect the per-request limits for each prompt.
 Really appreciate your sharp eye! 👀🔥

**Comment by bingshuailiu** [comment]
Hey @QsingHuan, check out this commit 👇
 🔗 [97da99e](https://github.com/ShopeeLLM/Spec-RL/commit/97da99ed54d663b238c417c728bc824c79cfab6f)
 The new change I mentioned above is included in this commit.

**Comment by QsingHuan** [comment]
Thanks a lot for your quick update!
Btw, I have another question about the saved rollout results. In the current implementation, the rollout is sorted then saved in the _dump_generations function. When using, align_prev_to_gen function is applied to search the matched result. Since data shuffle is already set to false in current setting, why bothering doing this instead of directly saving and loading without sorting? Thanks for your reply!

**Comment by bingshuailiu** [comment]
Hey @QsingHuan, great question!
I suspect that verl's default dump implementation might be changing the order of the saved results internally. In SPEC-RL verl 0.5.0 release version, there’s a re-balancing step that could affect data ordering across DP ranks:

```python
# Balance the number of valid tokens across DP ranks.
# NOTE: This usually changes the order of data in the `batch`,
# which won't affect the advantage calculation (since it's based on uid),
# but might affect the loss calculation (due to the change of mini-batching).
# TODO: Decouple the DP balancing and mini-batching.
if self.config.trainer.balance_batch:
    self._balance_batch(batch, metrics=metrics)
```

So, to be safe, I explicitly sort the rollout results before saving and then re-align them when loading.
This design helps avoid potential mismatches caused by hidden reordering inside verl.
The extra cost is minimal — sorting and alignment only add about 2 seconds per step, which is negligible compared to the overall training time.

**Comment by bingshuailiu** [comment]
> Thanks a lot for your quick update! Btw, I have another question about the saved rollout results. In the current implementation, the rollout is sorted then saved in the _dump_generations function. When using, align_prev_to_gen function is applied to search the matched result. Since data shuffle is already set to false in current setting, why bothering doing this instead of directly saving and loading without sorting? Thanks for your reply!

By the way, after thinking about your idea, I guess there’s probably a better way. We could try to capture the original rollout order at the beginning of each step (from gen_batch), and then save the dump following that order.
That way, the dumped results would stay exactly consistent with the later training process.

**Comment by QsingHuan** [comment]
I see, that is reasonable. 
However, I think maybe it is more convenient to save and load the prompt as well. And, from the 2nd epoch, we can just load data from the saved pt file, thus the dataloader is useless from then, and we don't need to do matching at all.

**Comment by QsingHuan** [comment]
> > Thanks a lot for your quick update! Btw, I have another question about the saved rollout results. In the current implementation, the rollout is sorted then saved in the _dump_generations function. When using, align_prev_to_gen function is applied to search the matched result. Since data shuffle is already set to false in current setting, why bothering doing this instead of directly saving and loading without sorting? Thanks for your reply!
> 
> By the way, after thinking about your idea, I guess there’s probably a better way. We could try to capture the original rollout order at the beginning of each step (from gen_batch), and then save the dump following that order. That way, the dumped results would stay exactly consistent with the later training process.

I think this is useful. Since current implementation has a limitation that we can't shuffle data between epochs. I wondering if this will cause damage to the training result since most of other algos do shuffling. With this idea plus an id attached to the data item, we can track the data globally, and data shuffle is available.

**Comment by bingshuailiu** [comment]
Hey @QsingHuan, Good point!
We actually compared runs with and without data.shuffle, and the results were almost identical — it doesn’t have a noticeable impact on training.
 spec-rl’s current implementation does support shuffle=True, but I haven’t focused much on it yet.
Right now, the shuffle behavior comes from verl’s built-in data loader logic — it only performs the shuffle once at the very beginning (as long as training isn’t interrupted).
 So after that initial shuffle, all later epochs will follow the same order — for example, if the first shuffle produces an order like [1, 3, 4, 6, 5, 2], then epoch 1, 2, 3, … will all use that same order.
Below is the relevant part from verl 0.5.0, which shows how the sampler is created when shuffle=True:
```python
import torch
from torch.utils.data import RandomSampler, SequentialSampler

# Use a sampler to facilitate checkpoint resumption.
# If shuffling is enabled in the data configuration, create a random sampler.
elif data_config.shuffle:
    train_dataloader_generator = torch.Generator()
    train_dataloader_generator.manual_seed(data_config.get("seed", 1))
    sampler = RandomSampler(data_source=dataset, generator=train_dataloader_generator)
else:
    # If shuffling is disabled, use a sequential sampler to iterate through the dataset in order.
    sampler = SequentialSampler(data_source=dataset)
```

So in short, this design means the dataset order is fixed after the first shuffle (seed-based reproducibility).
 We’ll also make sure to extend spec-rl’s speculative decoding and old-policy reuse to fully support data shuffle in future updates.

**Comment by QsingHuan** [comment]
Thanks so much for your reply. 
But I doubt if this statement is true. I think that the seed can fix the run-to-run order (the reproducibility you've mentioned), but the order between different epochs is still different. 
Consider the following example
```python
import torch
from torch.utils.data import Dataset, RandomSampler

class DemoDataset(Dataset):
    def __len__(self): 
        return 5
    def __getitem__(self, idx):
        return idx

train_dataloader_generator = torch.Generator()
train_dataloader_generator.manual_seed(42)
dataset = DemoDataset()
sampler = RandomSampler(dataset, generator=train_dataloader_generator)

print("Epoch 0:", list(sampler))
print("Epoch 1:", list(sampler))
```
will give result 
Epoch 0: [2, 4, 3, 0, 1]
Epoch 1: [0, 2, 3, 1, 4]
I'm not sure if this is actually the case in veRL or it some how set this.

**Comment by bingshuailiu** [comment]
I think you’re right, thanks for the detailed example! 
Looks like I mixed that up earlier, really appreciate your careful check and clarification.
I’ll take a closer look and work on adding proper data shuffle support for SPEC-RL as we discussed above.
Thanks again for such a thoughtful and in-depth discussion, it’s been super helpful! 💪

**Comment by bingshuailiu** [comment]
By the way, if you have any good implementation ideas, feel free to submit a PR — would love to collaborate on improving this together! 🚀

**Comment by bingshuailiu** [comment]
Since this issue was mainly about **“Questions about `per_request_max_new_tokens`”**, and we’ve already cleared that part, I’ll go ahead and close this issue for now.

Thanks again for the great discussion! 🙏
